src/rpc/core_rpc_server.cpp

```diff
@@ -1111,9 +1217,71 @@ namespace cryptonote
     return true;
   }
   //------------------------------------------------------------------------------------------------------------------------------
+  bool core_rpc_server::should_use_bootstrap_daemon()
+  {
+    if (m_bootstrap_daemon_address.empty())
+      return false;
+    if (!m_should_use_bootstrap_daemon)
+    {
+      MINFO("The local daemon is fully synced. Not switching back to the bootstrap daemon");
+      return false;
+    }
+    auto current_time = std::chrono::system_clock::now();
+    if (current_time - m_bootstrap_height_check_time > std::chrono::seconds(30))  // update every 30s
+    {
+      m_bootstrap_height_check_time = current_time;
+
+      uint64_t top_height;
+      crypto::hash top_hash;
+      m_core.get_blockchain_top(top_height, top_hash);
+      ++top_height; // turn top block height into blockchain height
+      uint64_t target_height = m_core.get_target_blockchain_height();
+
+      m_should_use_bootstrap_daemon = top_height < target_height;
+      MINFO((m_should_use_bootstrap_daemon ? "Using" : "Not using") << " the bootstrap daemon (top height: " << top_height << ", target height: " << target_height);
+    }
+    return m_should_use_bootstrap_daemon;
+  }
+  //------------------------------------------------------------------------------------------------------------------------------
+  template <typename COMMAND_TYPE>
+  bool core_rpc_server::use_bootstrap_daemon(const std::string &command_name, const typename COMMAND_TYPE::request& req, typename COMMAND_TYPE::response& res)
+  {
+    boost::lock_guard<boost::mutex> lock(m_bootstrap_daemon_mutex);
```

Comment:
Doesn't this make the daemon extremely slow and unresponsive?
It also allows the remote node to DOS your daemon.

Reply:
No, this is the same mutex lock used in every RPC. This bootstrap mode assumes that the daemon is only serving wallets in the same LAN; i.e. the daemon isn't supposed to accept connections from outside.

Reply:
Either you didn't get my point or I didn't understand the code ;)
When using the new `--bootstrap-daemon-address` CLI arg and the local blockchain is too far behind, your daemon acts as a proxy between your wallet (which is connected to your daemon) and the remote node. Doesn't this allow the remote node to DOS your daemon (and therefore all clients connected to it) if that remote node simply doesn't respond to requests originating from your deamon?

Reply:
The remote daemon can’t DOS your local daemon unless you open your port to the outside world and specify `—confirm-external-bind`. If the remote daemon decides not to respond, that’s just a useless remote node and you switch to a different one. It’s exactly like how the wallet’s `—daemon-address` flag works.

Reply:
> The remote daemon can’t DOS your local daemon

Yes it can (by not answering) and you don't need to open a port or specify some flag as _your daemon is the one connecting_ to the remote node. I just wanted to make it clear that if we add support to use more than one bootstrapping node in the future using some kind of connection pool, this could be problematic.

Anyway, ignore what I just said. What I originally intended to ask: What's the purpose of `m_bootstrap_daemon_mutex`? Why can't there be concurrent RPC requests?

Reply:
> Yes it can (by not answering) and you don't need to open a port or specify some flag as *your daemon is the one connecting* to the remote node.

I don't understand. If the remote node doesn't answer, the RPC request simply times out. This may cause a little bit of delay, and the user will quickly discover that the remote node isn't working as expected, so he'll switch to a different remote node.

> Anyway, ignore what I just said. What I originally intended to ask: What's the purpose of m_bootstrap_daemon_mutex? Why can't there be concurrent RPC requests?

I just followed the same practice used in the other parts of the codebase where HTTP request is handled (e.g. wallet2, simplewallet), and I just assume this is a requirement. Perhaps @moneromooo-monero can comment?

Reply:
IIRC, long ago, I think such locking was needed. Whether it is now, I do not know. Ideally the http code should be thread safe, but you'd have to read the code to make sure.

Reply:
Hmm, actually the mutex lock is used only in wallet2/simplewallet but not in wallet_rpc_server, so maybe this isn't a requirement here. Shall I remove it?

Reply:
simplewallet creates a refresh thread, while wallet_rpc_server does not.

Reply:
OK, I'll leave it as is then.